From 7cf770073434aba969788c68a46cc7f9dde51bc6 Mon Sep 17 00:00:00 2001
From: zrj <rimvydas.jasinskas@gmail.com>
Date: Wed, 28 Aug 2019 09:12:05 +0300
Subject: [PATCH 63/66] <signal.h>: Further reduce pthread namespace pollution.

 Only provide types mandated by POSIX here.
 Adjust thread libraries to always take types from _pthreadtypes.h.
---
 include/signal.h                   | 6 +++++-
 lib/libc_r/uthread/uthread_kill.c  | 1 +
 lib/libthread_xu/thread/thr_kill.c | 1 +
 sys/sys/_pthreadtypes.h            | 6 ++++++
 sys/sys/signal.h                   | 6 +++++-
 5 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/include/signal.h b/include/signal.h
index 51793abc36..6702d74a79 100644
--- a/include/signal.h
+++ b/include/signal.h
@@ -50,7 +50,11 @@ int	raise(int);
 #if __POSIX_VISIBLE || __XSI_VISIBLE
 int	kill(pid_t, int);
 #if __POSIX_VISIBLE >= 199506 || __XSI_VISIBLE >= 500
-#include <sys/_pthreadtypes.h>	/* XXX */
+#ifndef _PTHREAD_T_DECLARED
+#define _PTHREAD_T_DECLARED
+struct __pthread_s;
+typedef struct	__pthread_s		*pthread_t;
+#endif
 int	pthread_kill(pthread_t, int);
 int	pthread_sigmask(int, const sigset_t * __restrict,
 	    sigset_t * __restrict);
diff --git a/lib/libc_r/uthread/uthread_kill.c b/lib/libc_r/uthread/uthread_kill.c
index 7b00ac40ec..5b672a5741 100644
--- a/lib/libc_r/uthread/uthread_kill.c
+++ b/lib/libc_r/uthread/uthread_kill.c
@@ -32,6 +32,7 @@
  * $FreeBSD: src/lib/libc_r/uthread/uthread_kill.c,v 1.10.2.1 2002/10/22 14:44:03 fjoe Exp $
  */
 #include "namespace.h"
+#include <sys/_pthreadtypes.h>		/* for pthread_t */
 #include <errno.h>
 #include <signal.h>
 #include <pthread.h>
diff --git a/lib/libthread_xu/thread/thr_kill.c b/lib/libthread_xu/thread/thr_kill.c
index b178a443fe..70a47b8269 100644
--- a/lib/libthread_xu/thread/thr_kill.c
+++ b/lib/libthread_xu/thread/thr_kill.c
@@ -29,6 +29,7 @@
  */
 
 #include "namespace.h"
+#include <sys/_pthreadtypes.h>		/* for pthread_t */
 #include <machine/tls.h>
 #include <errno.h>
 #include <signal.h>
diff --git a/sys/sys/_pthreadtypes.h b/sys/sys/_pthreadtypes.h
index 8d11ed1b38..f57ebec190 100644
--- a/sys/sys/_pthreadtypes.h
+++ b/sys/sys/_pthreadtypes.h
@@ -68,8 +68,14 @@ struct __pthread_spinlock_s;
 /*
  * Basic types to be used in function prototypes.
  */
+#ifndef _PTHREAD_T_DECLARED
+#define _PTHREAD_T_DECLARED
 typedef struct	__pthread_s		*pthread_t;
+#endif
+#ifndef _PTHREAD_ATTR_T_DECLARED
+#define _PTHREAD_ATTR_T_DECLARED
 typedef struct	__pthread_attr_s	*pthread_attr_t;
+#endif
 typedef struct	__pthread_barrier_s	*pthread_barrier_t;
 typedef struct	__pthread_barrierattr_s	*pthread_barrierattr_t;
 typedef struct	__pthread_cond_s	*pthread_cond_t;
diff --git a/sys/sys/signal.h b/sys/sys/signal.h
index 630b922333..40d9de778b 100644
--- a/sys/sys/signal.h
+++ b/sys/sys/signal.h
@@ -238,7 +238,11 @@ union sigval {
 
 #if __POSIX_VISIBLE >= 199309 || __XSI_VISIBLE >= 500
 #if !defined(_KERNEL)
-#include <sys/_pthreadtypes.h>
+#ifndef _PTHREAD_ATTR_T_DECLARED
+#define _PTHREAD_ATTR_T_DECLARED
+struct __pthread_attr_s;
+typedef struct	__pthread_attr_s	*pthread_attr_t;
+#endif
 #endif
 
 struct sigevent {
-- 
2.22.0

